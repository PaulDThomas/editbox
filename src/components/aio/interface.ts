/** Individual options */
export interface AioOption {
  optionName?: AioNewItem;
  type: AioOptionType;
  value: unknown;
  label?: string;
  availableValues?: string[];
  readOnly?: boolean;
}

/** Text replacements, and their replacement matrix */
export interface AioReplacement<T> {
  airid?: string;
  oldText: string;
  newTexts: AioReplacementValues<T>[];
  includeTrailing?: boolean;
  externalName?: string;
}

export interface AioReplacementValues<T> {
  airid?: string;
  texts: T[];
  spaceAfter?: boolean;
  subLists?: AioReplacement<T>[];
}

export interface AioExternalReplacements<T> {
  givenName: string;
  newTexts: AioReplacementValues<T>[];
}

export interface AioExternalSingle<T> {
  airid: string;
  oldText: string;
  newText: T;
}

export enum AioOptionType {
  string = "string",
  number = "number",
  array = "array",
  object = "object",
  boolean = "boolean",
  select = "select",
  replacements = "replacements",
}

export enum AioNewItem {
  newKey = "newKey",
  newType = "newType",
}
