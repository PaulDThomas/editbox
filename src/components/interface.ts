import { EditorV3Style } from "@asup/editor-v3";

export interface AieStyleMap {
  [styleName: string]: { css: EditorV3Style; aieExclude?: string[] };
}
export interface AsupInternalEditorProps<T> {
  id: string;
  value?: T;
  setValue?: (ret: T) => void;
  style?: React.CSSProperties;
  styleMap?: AieStyleMap;
  textAlignment?: Draft.DraftComponent.Base.DraftTextAlignment | "decimal" | "default";
  decimalAlignPercent?: number;
  showStyleButtons?: boolean;
  editable?: boolean;
  className?: string;
  resize?: boolean;
  [key: string]: unknown;
}

export interface EditorProps<T> {
  Editor: (props: AsupInternalEditorProps<T>) => JSX.Element;
  getTextFromT: (text: T) => string[];
  replaceTextInT: (s: T, oldPhrase: string, newPhrase: T) => T;
  blankT: T;
  joinTintoBlock: (lines: T[]) => T;
  splitTintoLines: (text: T) => T[];
}
