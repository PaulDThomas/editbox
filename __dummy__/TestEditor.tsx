import { useEffect, useState } from "react";
import { AsupInternalEditorProps, EditorProps } from "../src/main";

export const TestEditor = <T,>(props: AsupInternalEditorProps<T>) => {
  const [text, setText] = useState<string>();
  useEffect(() => {
    if (typeof props.value === "string") setText(props.value);
  }, [props.value]);

  if (typeof props.value !== "string")
    throw new Error("If newText is not a string, a custom function is required");

  return (
    <input
      id={props.id}
      className={"aio-input"}
      disabled={!props.setValue}
      value={text ?? ""}
      onChange={(e) => setText(e.currentTarget.value)}
      onBlur={() => props.setValue && props.setValue(text as T)}
    />
  );
};

export const replaceTextInTestEditor = (text: string, oldPhrase: string, newPhrase: string) => {
  return text.replaceAll(oldPhrase, newPhrase);
};

export const testEditorProps: EditorProps<string> = {
  Editor: TestEditor,
  getTextFromT: (text: string) => [text],
  replaceTextInT: replaceTextInTestEditor,
  blankT: "",
  joinTintoBlock: (lines: string[]) => lines.join("\n"),
  splitTintoLines: (text: string) => text.split("\n"),
};
