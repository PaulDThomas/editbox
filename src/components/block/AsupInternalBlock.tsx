import React, { Dispatch, SetStateAction } from "react";
import { AioExternalSingle } from "../aio";
import { AieStyleMap, EditorProps } from "../interface";
import { AibBlock } from "./AibBlock";
import { BlockContextProvider } from "./BlockContentProvider";
import { AibBlockLine, AibLineType } from "./interface";

interface AsupInternalBlockProps<T> {
  id: string;
  lines: AibBlockLine<T>[];
  setLines?: Dispatch<SetStateAction<AibBlockLine<T>[]>>;
  disabled?: boolean;
  minLines?: number;
  maxLines?: number;
  externalSingles?: AioExternalSingle<T>[];
  styleMap?: AieStyleMap;
  defaultType?: AibLineType;
  canChangeType?: boolean;
  lineStyle?: React.CSSProperties;
  editorProps: EditorProps<T>;
}
export const AsupInternalBlock = <T,>({
  id,
  lines,
  setLines,
  disabled = false,
  minLines = 1,
  maxLines = 10,
  externalSingles = [],
  styleMap = {},
  defaultType = AibLineType.centerOnly,
  canChangeType = false,
  lineStyle,
  editorProps,
}: AsupInternalBlockProps<T>): JSX.Element => {
  return (
    <BlockContextProvider<T>
      initialState={{
        id,
        lines,
        returnData: (ret) => {
          setLines && setLines(ret);
        },
        minLines,
        maxLines,
        externalSingles,
        lineStyle,
        styleMap,
        defaultType,
        canChangeType,
        editorProps,
        disabled,
      }}
    >
      <AibBlock />
    </BlockContextProvider>
  );
};

AsupInternalBlock.displayName = "AsupInternalBlock";
