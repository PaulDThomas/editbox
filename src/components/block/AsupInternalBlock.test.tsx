import { ContextWindowStack } from "@asup/context-menu";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { act } from "react";
import { testEditorProps } from "../../../__dummy__/TestEditor";
import { AsupInternalBlock } from "./AsupInternalBlock";
import { AibBlockLine, AibLineType } from "./interface";

describe("AsupInternalBlock tests", () => {
  test("Basic render", async () => {
    const id = "line-id";
    const lines: AibBlockLine<string>[] = [
      {
        aifid: "1",
        lineType: AibLineType.leftAndRight,
        left: "First line",
        center: null,
        right: "Right text",
        canEdit: true,
        canRemove: true,
        canMove: true,
        addBelow: true,
        canChangeType: false,
      },
      {
        aifid: "2",
        lineType: AibLineType.centerOnly,
        left: null,
        center: "Second line",
        right: null,
        canEdit: true,
        canRemove: true,
        canMove: true,
        addBelow: true,
        canChangeType: false,
      },
    ];
    const setLines = jest.fn();
    const minLines = 1;
    const maxLines = 5;
    const user = userEvent.setup({ delay: null });

    await act(async () => {
      render(
        <ContextWindowStack>
          <div data-testid="container">
            <AsupInternalBlock
              id={id}
              lines={lines}
              setLines={setLines}
              minLines={minLines}
              maxLines={maxLines}
              editorProps={testEditorProps}
            />
          </div>
        </ContextWindowStack>,
      );
    });
    expect(screen.getByTestId("container")).toMatchSnapshot();

    // Add a line
    const addButton0 = screen.queryAllByLabelText("Add line")[0];
    expect(addButton0).toBeInTheDocument();
    await user.click(addButton0);

    expect(setLines).toHaveBeenLastCalledWith([
      lines[0],
      {
        aifid: "1001",
        lineType: AibLineType.centerOnly,
        left: null,
        center: "",
        right: null,
        canEdit: true,
        canMove: true,
        canRemove: true,
        addBelow: true,
        canChangeType: false,
      },
      lines[1],
    ]);
  });
});
