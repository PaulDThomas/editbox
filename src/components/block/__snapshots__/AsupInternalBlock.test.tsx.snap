// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AsupInternalBlock tests Basic render 1`] = `
<div
  data-testid="container"
>
  <div
    class="aibBlock"
    id="line-id"
  >
    <div
      class="aibLine"
      id="line-id-0"
    >
      <div
        class="aibLineButtons"
      />
      <div
        class="aiw-ContextWindow-module-RIIF9a-contextWindowAnchor"
      />
      <div
        class="aiw-ContextMenu-module-WinFxq-contextMenuHandler "
      >
        <div
          class="aibLineItemHolder"
        >
          <div
            class="aibLineItem  two"
          >
            <input
              class="aio-input"
              id="line-id-0-left-text"
              value="First line"
            />
          </div>
          <div
            class="aibLineItem  two"
          >
            <input
              class="aio-input"
              id="line-id-0-right-text"
              value="Right text"
            />
          </div>
        </div>
      </div>
      <div
        class="aibLineButtons"
      >
        <div
          class="aio-button-holder"
          id="line-id-0-add-line-holder"
        >
          <div
            class="aio-tip"
            style="display: flex; align-content: flex-center;"
          >
            <div
              aria-label="Add line"
              class="aiox-button aiox-plus"
              id="line-id-0-add-line"
              title="Add line"
            />
          </div>
        </div>
        <div
          class="aio-button-holder"
          id="line-id-0-remove-line-holder"
          style="color: red;"
        >
          <div
            class="aio-tip"
            style="display: flex; align-content: flex-center;"
          >
            <div
              aria-label="Remove line"
              class="aiox-button aiox-minus"
              id="line-id-0-remove-line"
              title="Remove line"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="aibLine"
      id="line-id-1"
    >
      <div
        class="aibLineButtons"
      />
      <div
        class="aiw-ContextWindow-module-RIIF9a-contextWindowAnchor"
      />
      <div
        class="aiw-ContextMenu-module-WinFxq-contextMenuHandler "
      >
        <div
          class="aibLineItemHolder"
        >
          <div
            class="aibLineItem  one"
            style="flex-grow: 1;"
          >
            <input
              class="aio-input"
              id="line-id-1-center-text"
              value="Second line"
            />
          </div>
        </div>
      </div>
      <div
        class="aibLineButtons"
      >
        <div
          class="aio-button-holder"
          id="line-id-1-add-line-holder"
        >
          <div
            class="aio-tip"
            style="display: flex; align-content: flex-center;"
          >
            <div
              aria-label="Add line"
              class="aiox-button aiox-plus"
              id="line-id-1-add-line"
              title="Add line"
            />
          </div>
        </div>
        <div
          class="aio-button-holder"
          id="line-id-1-remove-line-holder"
          style="color: red;"
        >
          <div
            class="aio-tip"
            style="display: flex; align-content: flex-center;"
          >
            <div
              aria-label="Remove line"
              class="aiox-button aiox-minus"
              id="line-id-1-remove-line"
              title="Remove line"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
